src/connector/src/sink/mqtt.rs

```diff
@@ -239,24 +352,256 @@ impl Drop for MqttSinkWriter {
 struct MqttSinkPayloadWriter {
     // connection to mqtt, one per executor
     client: rumqttc::v5::AsyncClient,
-    config: MqttConfig,
+    topic: Option<String>,
     qos: QoS,
     retain: bool,
+    topic_index_path: Vec<usize>,
 }
 
-impl FormattedSink for MqttSinkPayloadWriter {
-    type K = Vec<u8>;
-    type V = Vec<u8>;
+impl MqttSinkPayloadWriter {}
+
+impl MqttSinkPayloadWriter {
+    async fn write_chunk(&mut self, chunk: StreamChunk, encoder: &RowEncoderWrapper) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+
+            let topic = match get_topic_from_index_path(
+                &self.topic_index_path,
+                self.topic.as_deref(),
+                &row,
+            ) {
+                Some(s) => s,
+                None => {
+                    tracing::error!("topic field not found in row, skipping: {:?}", row);
+                    return Ok(());
+                }
+            };
+
+            let v = encoder.encode(row)?;
 
-    async fn write_one(&mut self, _k: Option<Self::K>, v: Option<Self::V>) -> Result<()> {
-        match v {
-            Some(v) => self
-                .client
-                .publish(&self.config.common.topic, self.qos, self.retain, v)
+            self.client
+                .publish(topic, self.qos, self.retain, v)
                 .await
                 .context("mqtt sink error")
-                .map_err(SinkError::Mqtt),
-            None => Ok(()),
+                .map_err(SinkError::Mqtt)?;
         }
+
+        Ok(())
+    }
+}
+
+fn get_topic_from_index_path<'s>(
+    path: &[usize],
+    default_topic: Option<&'s str>,
+    row: &'s RowRef<'s>,
```

Comment:
`RowRef` is already a reference type that implements `Copy`. We can pass `RowRef<'a>`.

Reply:
This is to avoid borrowing when using the datum_at function inside a closure and returning a value owned by the closure

Reply:
@wenym1 Do I need to do anything else?